File: services/mirror-worker/src/replay/order-created.ts
import axios from 'axios';
import { PoolClient } from 'pg';
import { config } from '../config';
import { log } from '../log';
import { ProdUserLog, prodSelectOne } from '../db/prod';
import {
  insertMapping,
  lookupMapping,
  stagingSelectOne,
  withStagingTx,
} from '../db/staging';
import { getStagingApiToken } from '../runtime';
import {
  translateAreaId,
  translateEquipmentId,
  translateSiteId,
} from '../translate';

// order-created creates a PO without starting it (status stays as 1=available).
// Used by CS Admin / frontend when batching multiple POs ahead of time.
//
// Sample prod payload:
//   {
//     idArea, idSite,
//     idOrder: "<biz key>",
//     idEquipment, idEnterprise,
//     equipmentSetup: [{ id, position }],
//     unitMultiplier,
//     productionOrderQuantity: "<number-as-string>"
//   }
//
// Output: POST /api/production-orders/create (CreateProductionOrderDto)
//   { idEnterprise, idSite, idArea, idEquipment, idOrder, productionOrderQuantity }
//
// Idempotency: payload doesn't carry prod's id_production_order, so we resolve
// it after the fact by querying prod for the PO matching (idEnterprise, idOrder)
// created around row.ts_event. If mapping already exists, skip.
interface OrderCreatedPayload {
  idArea: number;
  idSite: number;
  idOrder: number | string;
  idEquipment: number;
  idEnterprise: number;
  productionOrderQuantity: number | string;
}

export async function replayOrderCreated(
  _client: PoolClient,
  row: ProdUserLog,
): Promise<void> {
  const payload = row.payload as OrderCreatedPayload;
  if (!payload || typeof payload !== 'object') {
    throw new Error('order-created payload missing or not an object');
  }
  if (typeof payload.idEquipment !== 'number') {
    throw new Error('order-created payload missing idEquipment');
  }
  if (payload.idOrder === undefined || payload.idOrder === null) {
    throw new Error('order-created payload missing idOrder');
  }

  const idOrderNum = Number(payload.idOrder);

  const stagingEquipmentId = await translateEquipmentId(payload.idEquipment);
  const stagingSiteId = await translateSiteId(payload.idSite);
  const stagingAreaId = await translateAreaId(payload.idArea);

  // Resolve prod's surrogate id_production_order for this row so we can map it.
  // The user_logs row was written right after edge-api created the PO, so the
  // most recent prod PO matching (id_enterprise, id_order) is ours.
  const prodPoRow = await prodSelectOne<{ id_production_order: string }>(
    `SELECT id_production_order::text
       FROM production_orders
      WHERE id_enterprise = $1 AND id_order = $2
      ORDER BY id_production_order DESC LIMIT 1`,
    [config.prodEnterpriseId, idOrderNum],
  );
  if (!prodPoRow) {
    throw new Error(
      `no prod production_order found for idEnterprise=${config.prodEnterpriseId} idOrder=${idOrderNum}`,
    );
  }
  const prodPoId = BigInt(prodPoRow.id_production_order);

  // Already mapped? Skip (idempotent re-run).
  const existing = await lookupMapping('production_order', config.sourceName, prodPoId);
  if (existing !== undefined) {
    log.info(
      { sourceLogId: row.id_user_logs, prodPoId: prodPoId.toString(), stagingPoId: existing.toString() },
      'order-created already mapped, skipping',
    );
    return;
  }

  const body = {
    idEnterprise: config.stagingEnterpriseId,
    idSite: stagingSiteId,
    idArea: stagingAreaId,
    idEquipment: stagingEquipmentId,
    idOrder: idOrderNum,
    productionOrderQuantity: Number(payload.productionOrderQuantity),
  };

  const res = await axios.post(
    `${config.stagingApi.baseUrl}/api/production-orders/create`,
    body,
    {
      params: {
        token: getStagingApiToken(),
        idEnterprise: config.stagingEnterpriseId,
      },
      headers: {
        'x-user': row.nm_user ?? 'mirror-worker',
        'X-Mirror-Source': config.sourceName,
      },
      validateStatus: () => true,
      timeout: 15_000,
    },
  );
  if (res.status >= 400) {
    throw new Error(
      `staging /api/production-orders/create returned ${res.status}: ${JSON.stringify(res.data)}`,
    );
  }

  // Look up the new staging PO by business key and record mapping.
  const created = await stagingSelectOne<{ id_production_order: string }>(
    `SELECT id_production_order::text
       FROM production_orders
      WHERE id_enterprise = $1 AND id_order = $2
      ORDER BY id_production_order DESC LIMIT 1`,
    [config.stagingEnterpriseId, idOrderNum],
  );
  if (!created) {
    throw new Error(
      `create returned ${res.status} but staging PO with idOrder=${idOrderNum} not found`,
    );
  }
  const stagingPoId = BigInt(created.id_production_order);
  await withStagingTx(async (c) => {
    await insertMapping(c, {
      entityType: 'production_order',
      source: config.sourceName,
      prodId: prodPoId,
      stagingId: stagingPoId,
      sourceLogId: BigInt(row.id_user_logs),
    });
  });

  log.info(
    {
      sourceLogId: row.id_user_logs,
      prodPoId: prodPoId.toString(),
      stagingPoId: stagingPoId.toString(),
      idOrder: idOrderNum,
      status: res.status,
    },
    'replayed order-created',
  );
}

File: services/mirror-worker-go/internal/replay/order_created.go
package replay

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"

	"github.com/jackc/pgx/v5"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/config"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/db"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/translate"
)

// OrderCreatedPayload — prod payload is wider than these fields (also
// equipmentSetup, unitMultiplier, idEnterprise); we only need what staging's
// CreateProductionOrderDto wants. idOrder is `number | string` in prod —
// json.Number handles both without losing precision.
type OrderCreatedPayload struct {
	IDArea                  int         `json:"idArea"`
	IDSite                  int         `json:"idSite"`
	IDOrder                 json.Number `json:"idOrder"`
	IDEquipment             int         `json:"idEquipment"`
	ProductionOrderQuantity json.Number `json:"productionOrderQuantity"`
}

// OrderCreated: replay /api/production-orders/create then resolve the
// resulting prod PO + staging PO and persist the mapping. Same shape as
// order-created-started, only differs by endpoint + body.
//
// Idempotency: prod payload doesn't carry id_production_order, so we look
// it up post-hoc via (id_enterprise, id_order, ts_creation >= ts_event - 1m).
// If a mapping row already exists for that prod PO id, skip.
func OrderCreated(
	cfg *config.Config,
	t *translate.Translator,
	prodDB *db.Prod,
	stagingDB *db.Staging,
	apiToken func() string,
	httpc *http.Client,
	logger *slog.Logger,
) Handler {
	return func(ctx context.Context, tx pgx.Tx, row db.ProdUserLog) error {
		var p OrderCreatedPayload
		if err := decodeWithNumbers(row.Payload, &p); err != nil {
			return fmt.Errorf("parse order-created payload: %w", err)
		}
		if p.IDEquipment == 0 || p.IDOrder == "" {
			return fmt.Errorf("order-created payload missing fields: %+v", p)
		}
		idOrderNum, err := p.IDOrder.Int64()
		if err != nil {
			return fmt.Errorf("idOrder not numeric: %w", err)
		}
		qty, err := p.ProductionOrderQuantity.Float64()
		if err != nil {
			return fmt.Errorf("productionOrderQuantity not numeric: %w", err)
		}

		stagingEqID, err := t.Equipment(ctx, p.IDEquipment)
		if err != nil {
			return fmt.Errorf("translate equipment: %w", err)
		}
		stagingSiteID, err := t.Site(ctx, p.IDSite)
		if err != nil {
			return fmt.Errorf("translate site: %w", err)
		}
		stagingAreaID, err := t.Area(ctx, p.IDArea)
		if err != nil {
			return fmt.Errorf("translate area: %w", err)
		}

		// Find the prod PO we're mirroring — most recent one matching
		// (enterprise, id_order) created near this user_logs row.
		var prodPOIDStr string
		// ts_creation: prod's production_orders has no created_at column
		// (SQLSTATE 42703 in DLQ id 284).
		found, err := prodDB.SelectOne(ctx,
			`SELECT id_production_order::text FROM production_orders
			  WHERE id_enterprise = $1 AND id_order = $2
			    AND ts_creation >= $3::timestamptz - interval '1 minute'
			  ORDER BY id_production_order DESC LIMIT 1`,
			[]any{cfg.ProdEnterpriseID, idOrderNum, row.TsEvent}, &prodPOIDStr)
		if err != nil {
			return fmt.Errorf("prod PO lookup: %w", err)
		}
		if !found {
			return fmt.Errorf("no prod production_order for id_enterprise=%d id_order=%d", cfg.ProdEnterpriseID, idOrderNum)
		}
		prodPOID, err := parseBigint(prodPOIDStr)
		if err != nil {
			return fmt.Errorf("parse prod PO id: %w", err)
		}

		// Already mapped? Skip — idempotent re-run.
		if existing, hit, err := stagingDB.LookupMapping(ctx, "production_order", cfg.SourceName, prodPOID); err != nil {
			return fmt.Errorf("mapping lookup: %w", err)
		} else if hit {
			logger.Info("order-created already mapped, skipping",
				slog.Int64("sourceLogID", row.IDUserLogs),
				slog.Int64("prodPOID", prodPOID),
				slog.Int64("stagingPOID", existing))
			return nil
		}

		body := map[string]any{
			"idEnterprise":            cfg.StagingEnterpriseID,
			"idSite":                  stagingSiteID,
			"idArea":                  stagingAreaID,
			"idEquipment":             stagingEqID,
			"idOrder":                 idOrderNum,
			"productionOrderQuantity": qty,
		}
		status, _, err := PostStaging(ctx, cfg, httpc, apiToken(), row,
			"/api/production-orders/create", body)
		if err != nil {
			return err
		}

		// Resolve the staging PO just created.
		var stagingPOIDStr string
		found, err = stagingDB.SelectOne(ctx,
			`SELECT id_production_order::text FROM production_orders
			  WHERE id_enterprise = $1 AND id_order = $2
			  ORDER BY id_production_order DESC LIMIT 1`,
			[]any{cfg.StagingEnterpriseID, idOrderNum}, &stagingPOIDStr)
		if err != nil {
			return fmt.Errorf("staging PO lookup: %w", err)
		}
		if !found {
			return fmt.Errorf("create returned %d but staging PO with id_order=%d not found", status, idOrderNum)
		}
		stagingPOID, err := parseBigint(stagingPOIDStr)
		if err != nil {
			return fmt.Errorf("parse staging PO id: %w", err)
		}

		// Insert via the per-row tx — atomic with cursor advance done by
		// processRow. The TS code originally opened a SECOND tx here
		// (see 7e5b0bf for the bug it caused); the Go port skips that
		// trap by reusing the outer tx.
		if err := db.InsertMapping(ctx, tx, db.MapInsert{
			EntityType:  "production_order",
			Source:      cfg.SourceName,
			ProdID:      prodPOID,
			StagingID:   stagingPOID,
			SourceLogID: row.IDUserLogs,
		}); err != nil {
			return fmt.Errorf("insert mapping: %w", err)
		}

		logger.Info("replayed order-created",
			slog.Int64("sourceLogID", row.IDUserLogs),
			slog.Int64("prodPOID", prodPOID),
			slog.Int64("stagingPOID", stagingPOID),
			slog.Int64("idOrder", idOrderNum),
			slog.Int("status", status))
		return nil
	}
}
